File: bindings/java-jni/src/main/java/org/maplibre/nativejni/map/TileOptions.java
package org.maplibre.nativejni.map;

import java.util.Objects;

/** Mutable descriptor for tile prefetch and level-of-detail controls. */
public final class TileOptions {
  private Integer prefetchZoomDelta;
  private Double lodMinRadius;
  private Double lodScale;
  private Double lodPitchThreshold;
  private Double lodZoomShift;
  private TileLodMode lodMode;

  public boolean hasPrefetchZoomDelta() {
    return prefetchZoomDelta != null;
  }

  public Integer prefetchZoomDelta() {
    return prefetchZoomDelta;
  }

  public TileOptions prefetchZoomDelta(int prefetchZoomDelta) {
    this.prefetchZoomDelta = prefetchZoomDelta;
    return this;
  }

  public TileOptions clearPrefetchZoomDelta() {
    prefetchZoomDelta = null;
    return this;
  }

  public boolean hasLodMinRadius() {
    return lodMinRadius != null;
  }

  public Double lodMinRadius() {
    return lodMinRadius;
  }

  public TileOptions lodMinRadius(double lodMinRadius) {
    this.lodMinRadius = lodMinRadius;
    return this;
  }

  public TileOptions clearLodMinRadius() {
    lodMinRadius = null;
    return this;
  }

  public boolean hasLodScale() {
    return lodScale != null;
  }

  public Double lodScale() {
    return lodScale;
  }

  public TileOptions lodScale(double lodScale) {
    this.lodScale = lodScale;
    return this;
  }

  public TileOptions clearLodScale() {
    lodScale = null;
    return this;
  }

  public boolean hasLodPitchThreshold() {
    return lodPitchThreshold != null;
  }

  public Double lodPitchThreshold() {
    return lodPitchThreshold;
  }

  public TileOptions lodPitchThreshold(double lodPitchThreshold) {
    this.lodPitchThreshold = lodPitchThreshold;
    return this;
  }

  public TileOptions clearLodPitchThreshold() {
    lodPitchThreshold = null;
    return this;
  }

  public boolean hasLodZoomShift() {
    return lodZoomShift != null;
  }

  public Double lodZoomShift() {
    return lodZoomShift;
  }

  public TileOptions lodZoomShift(double lodZoomShift) {
    this.lodZoomShift = lodZoomShift;
    return this;
  }

  public TileOptions clearLodZoomShift() {
    lodZoomShift = null;
    return this;
  }

  public boolean hasLodMode() {
    return lodMode != null;
  }

  public TileLodMode lodMode() {
    return lodMode;
  }

  public TileOptions lodMode(TileLodMode lodMode) {
    var value = Objects.requireNonNull(lodMode, "lodMode");
    value.nativeValue();
    this.lodMode = value;
    return this;
  }

  public TileOptions clearLodMode() {
    lodMode = null;
    return this;
  }
}

File: bindings/kotlin-native/src/nativeMain/kotlin/org/maplibre/nativeffi/map/TileOptions.kt
package org.maplibre.nativeffi.map

/** Mutable descriptor for tile prefetch and level-of-detail controls. */
public class TileOptions {
  public var prefetchZoomDelta: Int? = null
    private set

  public var lodMinRadius: Double? = null
    private set

  public var lodScale: Double? = null
    private set

  public var lodPitchThreshold: Double? = null
    private set

  public var lodZoomShift: Double? = null
    private set

  public var lodMode: TileLodMode? = null
    private set

  public fun hasPrefetchZoomDelta(): Boolean = prefetchZoomDelta != null

  public fun prefetchZoomDelta(prefetchZoomDelta: Int): TileOptions {
    require(prefetchZoomDelta >= 0) { "prefetchZoomDelta must be non-negative" }
    this.prefetchZoomDelta = prefetchZoomDelta
    return this
  }

  public fun clearPrefetchZoomDelta(): TileOptions = apply { prefetchZoomDelta = null }

  public fun hasLodMinRadius(): Boolean = lodMinRadius != null

  public fun lodMinRadius(lodMinRadius: Double): TileOptions = apply {
    this.lodMinRadius = lodMinRadius
  }

  public fun clearLodMinRadius(): TileOptions = apply { lodMinRadius = null }

  public fun hasLodScale(): Boolean = lodScale != null

  public fun lodScale(lodScale: Double): TileOptions = apply { this.lodScale = lodScale }

  public fun clearLodScale(): TileOptions = apply { lodScale = null }

  public fun hasLodPitchThreshold(): Boolean = lodPitchThreshold != null

  public fun lodPitchThreshold(lodPitchThreshold: Double): TileOptions = apply {
    this.lodPitchThreshold = lodPitchThreshold
  }

  public fun clearLodPitchThreshold(): TileOptions = apply { lodPitchThreshold = null }

  public fun hasLodZoomShift(): Boolean = lodZoomShift != null

  public fun lodZoomShift(lodZoomShift: Double): TileOptions = apply {
    this.lodZoomShift = lodZoomShift
  }

  public fun clearLodZoomShift(): TileOptions = apply { lodZoomShift = null }

  public fun hasLodMode(): Boolean = lodMode != null

  public fun lodMode(lodMode: TileLodMode): TileOptions = apply {
    require(lodMode != TileLodMode.UNKNOWN) { "lodMode must be a known value" }
    this.lodMode = lodMode
  }

  public fun clearLodMode(): TileOptions = apply { lodMode = null }
}
